### ISSUE TITLE
replace adamW and pagedadam with 8bitpagedadam or torchao CPUOffloadOptimizer

### ISSUE THREAD

**Issue by felipemello1** [pytorch/torchtune#1576]
Apparently there is no reason to use paged adam instead of the 8bit version. We could replace it.

Also, full finetune single device should use paged adam, instead of adamw, for better memory.

For single device, we have torchaos version that is faster than the one from bitsandbytes: https://github.com/pytorch/ao/blob/8236a874479a9a9168e584c81dda8707f4c41006/torchao/prototype/low_bit_optim/cpu_offload.py#L9

**Comment by SalmanMohammadi** [comment]
These are just for our full finetune low memory configs right? I almost wonder if we should re-benchmark this recipe with all the new memory optimizations that have been coming in.

**Comment by NeuralFlux** [comment]
> Apparently there is no reason to use paged adam instead of the 8bit version. We should replace it.
> 

@felipemello1 can you please cite the source? I'm deciding between optimizers too.

**Comment by felipemello1** [comment]
@NeuralFlux I dont have it :/ But what i heard from some other coworkers is that they didnt observe change in the loss. Are you doing full finetuning? If so, you need PagedAdam to save you memory.

If you are using LoRA, you dont need pagedAdam, since the gradients are not your bottleneck. You can just use AdamW with fused=True.

**Comment by NeuralFlux** [comment]
No worries! I'm doing QLoRA but keep running into OOM because of big sequence lengths. I'm using pagedAdam to save how much ever memory I can. I noticed we are not compatible with `torchao==0.5.0` yet, which has the CPU offloader. And what does `fused=True` do? I read the AdamW docs but could not make sense of it.

**Comment by felipemello1** [comment]
How big is the sequence length? Also, are you using torchtune nighlities? If you arent, please try this:

```
pip install --pre torch torchvision torchao --index-url https://download.pytorch.org/whl/nightly/cu121
pip install --pre torchtune --extra-index-url https://download.pytorch.org/whl/nightly/cpu --no-cache-dir
pip install --pre torchao --index-url https://download.pytorch.org/whl/nightly/cu121 
```

And run your model with compile=True

you should see a huge difference in memory/tokens per second

**Comment by felipemello1** [comment]
Make sure that your config is using the chunked cross entropy, like we have in our default configs

**Comment by NeuralFlux** [comment]
Sure! Also, I noticed [QLoRA config](https://github.com/pytorch/torchtune/blob/main/recipes/configs/llama3_1/8B_qlora_single_device.yaml) mentions dtype as `bfloat16`. Is this compute dtype or storage dtype?

I tried installing but pip still tells me
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torchtune 0.2.1 requires torchao==0.3.1, but you have torchao 0.5.0 which is incompatible.
```

**Comment by felipemello1** [comment]
hmm, maybe try a fresh environment?

conda create -n your_env_name python=3.10
conda activate your_env_name
pip install --pre torch torchvision torchao --index-url https://download.pytorch.org/whl/nightly/cu121
pip install --pre torchtune --extra-index-url https://download.pytorch.org/whl/nightly/cpu --no-cache-dir
pip install --pre torchao --index-url https://download.pytorch.org/whl/nightly/cu121 

> Is this compute dtype or storage dtype?
compute is done in bf16. The quantize layers that are not being trained are stored in NF4, which saves 15GiB -> 5GiB if your model is llama 8B

**Comment by NeuralFlux** [comment]
Gotcha, I will try that soon (weekend's about to start here haha). Have a good weekend!

**Comment by NeuralFlux** [comment]
Hi @felipemello1 . I launched a job over the weekend that worked. Setting `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` and using `bitsandbytes.optim.PagedAdamW` did the job. I think the former was the main issue because memory reserved by PyTorch was in GBs. Thanks a lot for your inputs!

**Comment by FurkanGozukara** [comment]
> `PagedAdamW`

do you know difference between PagedAdEMAMix8bit and bitsandbytes.optim.AdEMAMix8bit

**Comment by ebsmothers** [comment]
Hi @FurkanGozukara I'm not familiar with PagedAdEMAMix8bit, can you share a link?

**Comment by FurkanGozukara** [comment]
> Hi [@FurkanGozukara](https://github.com/FurkanGozukara) I'm not familiar with PagedAdEMAMix8bit, can you share a link?

https://github.com/kohya-ss/sd-scripts/pull/1640

**Comment by ebsmothers** [comment]
@FurkanGozukara thanks for sharing the link. Personally I'm not too familiar with these optimizers as we mostly use ones that are available in bitsandbytes. Maybe best to ask on the sd-scripts repo directly?

**Comment by FurkanGozukara** [comment]
> [@FurkanGozukara](https://github.com/FurkanGozukara) thanks for sharing the link. Personally I'm not too familiar with these optimizers as we mostly use ones that are available in bitsandbytes. Maybe best to ask on the sd-scripts repo directly?

thanks. paged seems like automatically using entire vram but not throwing out of vram really good :)

**Comment by ebsmothers** [comment]
Going to close this issue as it's more of a question for sd-scripts. @FurkanGozukara feel free to reopen if you need any more assistance on optimizer usage in torchtune
